### ISSUE TITLE
ICM7228C issues with Library

### ISSUE THREAD

**Issue by gyakubov** [Andy4495/ICM7218#4]
Hello Andy.  I am trying to use your library to work with ICM7228C chip connected to 5 positions 7 segments LED.  Unfortunately cannot make it work.  Hardware connected with Aquino Uno and I tested it without any library sending digits directly . It works fine so I am pretty sure that HW is OK.  When I am employing library it does not make any effect. 
Would you, please, suggest any troubleshooting steps.

**Comment by Andy4495** [comment]
Hi - 
I didn't have a 7228C to test with, so I'm hoping I got the code right from reading the data sheet, so this will give me a chance to get the library tested on that chip. 

To start, can you provide a wiring diagram (picture of a hand-drawn diagram is fine), showing how the Arduino and chip are connected to each other and powered. Also, list out the constructor that you are using (or the whole sketch if it isn't too long). 

After I see that, we'll see if we can figure out the next steps. 

Thanks,
Andy

**Comment by gyakubov** [comment]
Hi Andy.  Thanks for quick response. below is the relevant piece of code   -just a constructor and setup function.

`#include <ICM7218.h>

ICM7218 myLED(6, 7, 10, 12, 13, A1, 3, 4, 0, 5, 1);

void setup() {

  Serial.begin(9600);
  myLED.setMode(ICM7218::CODEB);
  myLED.print(0,0);
}

void loop() {
}
`
I was trying to print various digits in various positions with no any results

![ICM7228C_Arduino](https://github.com/Andy4495/ICM7218/assets/37917533/c1efa8f4-560a-492a-8175-9c0d7d677e92)

**Comment by Andy4495** [comment]
I took a closer look at the code and found a couple possible issues. In particular, I think the code flips DA0 and DA2, but since you are trying to print to position zero, that won't matter. I also think I might be handling the mode pin incorrectly which could put it into HEXA mode when you select CODEB.

I don't think either of these issues would cause you to see nothing on the display; at worst, I think they would cause the LED to display something, but possible the wrong character in the wrong position. 

So let's start with the wiring, and see if we can get something displayed. Once we know that is working, then I'll work on updating the code. 

As far as your wiring -- the schematic doesn't quite match your constructor. For ID0, DA1, DA2, and WRITE, it appears that you are using the chip pin number instead of the Arduino digital pin number. For example, on your schematic, ID0 is connected to chip pin 6 which corresponds to Arduino pin D4. So the first parameter in the constructor should be 4, not 6, if your schematic is correct. Other pins in the constructor (e.g. ID1, ID2, ID3) match the schematic.

Also, the Arduino uses pins D0 and D1 for the serial/USB port, so those should not be used to control the LED. The schematic shows you using D1, but the constructor does not. I'm not sure which is correct, but double check and use alternate pins instead of D0/D1. 

Finally, the schematic has the MODE pin (HEX/CODEB/SHUTDOWN) left floating. That is fine, and it hardwires the chip to CODEB mode. In the constructor, however, you should use the enum `ICM7218::NO_PIN`, not `0`.

One other thing, just to make sure we're testing this completely -- position zero in the `print(c, pos)` method refers to digit 7 on the ICM7228 chip -- so confirm that there is an LED connected to that output, or add 7 more print statements to your test code (one for each position 0 - 7) just to make sure. 

So, please confirm that the schematic matches the constructor, note the other comments, and let me know if that changes anything. 

Thanks.

**Comment by Andy4495** [comment]
Here is an updated library file which I think fix the two issues I mentioned above. 

Confirm that your constructor matches your wiring and try out this updated code and let me know if that fixes the problem. As mentioned above, I would suggest that you add more `print()` statements to your test code just in case there is something wrong with the digit addressing (in the code or wiring). 

ICM7218.cpp:
*EDIT 5-Dec-2023: Removed code listing since fix has been implemented on branch "issue_4" and will be merged to main and released.*

**Comment by Andy4495** [comment]
I was able to acquire some 7228D chips (same programming interface as 7228C). I updated the library and tested it with the 7228D chip and I think I have resolved the issue with the library for the C/D chips. I published a new release 1.3.2. When you get a chance, download a new version of the library and see if that fixes the problem. If it still doesn't work, open a new issue and I'll take a look. 

- Andy
